src/event.c

```diff
@@ -132,6 +167,35 @@ void io_set(io_t *io, int flags) {
 	}
 
 #ifndef HAVE_MINGW
+#ifdef HAVE_SYS_EPOLL_H
+	struct epoll_event ev;
+	ev.data.fd = io->fd;
+	ev.data.ptr = io;
+
+	epoll_ctl(epollset, EPOLL_CTL_DEL, io->fd, NULL);
+	io_tree.generation++;
+
+	if((flags & IO_READ) && (flags & IO_WRITE)) {
+		ev.events = EPOLLIN | EPOLLOUT;
+	}
+
+	else if(flags & IO_READ) {
+		ev.events = EPOLLIN;
+	}
+
+	else if(flags & IO_WRITE) {
+		ev.events = EPOLLOUT;
+	}
+
+	else {
+		return;
```

Comment:
It should be valid to call `epoll_ctl(EPOLL_CTL_ADD)` with `ev.events = 0`. Also, instead of doing `EPOLL_CTL_DEL` followed by `EPOLL_CTL_DEL`, why not use `EPOLL_CTL_MOD`?

Reply:
epoll mod won't add (ENOENT)
epoll add won't mod (EEXIST)

We don't know if the fd is being modified or added in the current interface so a DEL+ADD is required.

Reply:
There would be gains in keeping fds permanently registered in general operation for both read at a given time and then being able to just EPOLL_CTL_MOD reliably to remove write as appropriate.

Reply:
I agree, it would be nice if tinc followed epoll requirements.

Reply:
Do you want to do this in this PR or want to do it in a separate one? I wouldn't mind the less efficient DEL+ADD for now to avoid having to make too many changes in one go.

Reply:
Seperate for sure. Not certain the best way to do it yet.